go/kbnm/handler.go

```diff
@@ -53,25 +59,99 @@ func (h *handler) handleChat(req *Request) error {
 		return errMissingField
 	}
 
-	binPath, err := findKeybaseBinary()
+	binPath, err := h.FindKeybaseBinary()
 	if err != nil {
 		return err
 	}
 
+	var out bytes.Buffer
 	cmd := exec.Command(binPath, "chat", "send", "--private", req.To)
+	cmd.Env = append(os.Environ(), "KEYBASE_LOG_FORMAT=plain")
 	cmd.Stdin = strings.NewReader(req.Body)
+	cmd.Stdout = &out
+	cmd.Stderr = &out
 
-	// TODO: Check/convert status code more precisely? Maybe return stdout as
-	// part of the error if there is one?
-	err = h.Run(cmd)
+	if err := h.Run(cmd); err != nil {
+		return parseError(&out, err)
+	}
 
-	return err
+	return nil
 }
 
 type resultQuery struct {
-	Sigs []struct {
-		Statement string `json:"statement"`
-	} `json:"sigs"`
+	Username string `json:"username"`
+}
+
+// parseQuery reads the stderr from a keybase query command and returns a result
+func parseQuery(r io.Reader) (*resultQuery, error) {
+	scanner := bufio.NewScanner(r)
+
+	for scanner.Scan() {
+		// Find a line that looks like... "[INFO] 001 Identifying someuser"
+		line := strings.TrimSpace(scanner.Text())
+		parts := strings.Split(line, " ")
+		if len(parts) < 4 {
+			continue
+		}
+
+		if parts[2] != "Identifying" {
+			continue
+		}
+
+		resp := &resultQuery{
+			Username: parts[3],
+		}
+		return resp, nil
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, scanner.Err()
+	}
+
+	return nil, errUserNotFound
+}
+
+// parseError reads stderr output and returns an error made from it. If it
+// fails to parse an error, it returns the fallback error.
+func parseError(r io.Reader, fallback error) error {
+	scanner := bufio.NewScanner(r)
+
+	// Find the final error
+	var lastErr error
+	for scanner.Scan() {
+		// Should be of the form "[ERRO] 001 Not found" or "...: No resolution found"
+		line := strings.TrimSpace(scanner.Text())
+		if line == "" {
+			continue
+		}
+		// Check some error states we know about:
+		if strings.Contains(line, "Keybase isn't running.") {
+			return errKeybaseNotRunning
+		}
+		if strings.Contains(line, "You are not logged into Keybase.") {
+			return errKeybaseNotLoggedIn
+		}
+		parts := strings.SplitN(line, " ", 3)
+		if len(parts) < 3 {
+			continue
+		}
+		if parts[0] != "[ERRO]" {
+			continue
+		}
+		if strings.HasSuffix(parts[2], "No resolution found") {
+			return errUserNotFound
+		}
+		if strings.HasPrefix(parts[2], "Not found") {
+			return errUserNotFound
+		}
+		lastErr = fmt.Errorf(parts[2])
+	}
```

Comment:
We should probably throw some comments into client next to these strings, being like "hey if you're changing this, it'll probably break the extension :p"